### PR TITLE
Add `distributed` flag to VC to enable support for DVT

### PR DESCRIPTION
**Pull request by AgeManning**
Am interested in feedback for this PR. 

There are efforts towards building out Distributed Validator Technology (DVT). We want to support this effort and make Lighthouse maximally compatible with what is being built in the space. 

For an immediate (perhaps temporary) path forward, I'm suggesting we group some of the required functionality under a `--distributed` CLI flag, which is counter to some opinions expressed in https://github.com/sigp/lighthouse/issues/4248. 

I've added additional reasoning to the above issue, but in the current context, there is desire to add a minimal change to the VC which I think is fine to add behind either a CLI flag or feature flag. I expect more changes to occur and my thinking is we can use this banner flag to incorporate any future changes we need to support DVT under this CLI flag. 

For this PR the change is to compute sync aggregates for duties per slot, rather than to pre-compute them for each sync committee period.

### CODE REVIEW

File: validator_client/src/cli.rs [hunk 1]
```diff
@@ -153,6 +153,12 @@ pub fn cli_app<'a, 'b>() -> App<'a, 'b> {
                 .value_name("FEE-RECIPIENT")
                 .takes_value(true)
         )
+        .arg(
+            Arg::with_name("distributed")
+                .long("distributed")
+                .help("Enables functionality required for running the validator in a distributed network.")
```
Comment (xenowits) [inline validator_client/src/cli.rs]: nit
```suggestion
                .help("Enables functionality required for running the validator in a distributed validator cluster.")
```

**Comment by xenowits** [comment]
Thanks @AgeManning , looks good!

Seems calculating sync committee aggregations per slot is still in progress, happy to review it when implemented 🙌

**Comment by michaelsproul** [comment]
I've added a commit 8d196152a585f2930e8f0721b2534616fdeb2d98 to compute sync committee selection proofs with 1-slot lookahead when `--distributed` is provided.

The only change for non-distributed validator clients is that selection proofs will **no longer be computed in epoch-sized batches**. Rather each batch is now just a single slot. I don't think this will have a substantial performance impact (positive or negative), but will be deploying to one of our testnet nodes with 10k+ keys to make sure.

**Comment by michaelsproul** [comment]
Looks like this is working as intended. Proofs are computed at the start of the previous slot (7104128=06:25:36):

> Dec 05 06:25:36.353 DEBG Calculating sync selection proofs       pre_compute_slot: 7104129, current_slot: 7104128, period: 867, service: duties

A validator is found to be a sync aggregator (the slot here is the _proof slot_, which is the slot prior, i.e. the current slot):

> Dec 05 06:25:36.354 DEBG Validator is sync aggregator            subnet_id: 3, slot: 7104128, validator_index: 115668, service: duties

Then, 4s into the slot 7104129, the sync aggregate is successfully published:

> Dec 05 06:25:56.674 INFO Successfully published sync contributions, slot: 7104129, num_signers: 93, beacon_block_root: 0xeb8b…481e, subnet: 3, service: sync_committee

**Comment by michaelsproul** [comment]
@xenowits It would be great if you could take this PR for a spin with Obol and provide feedback :pray:

**Comment by xenowits** [comment]
Hey @michaelsproul , that's some great work!

Unfortunately, i'm out for the next two weeks. I'll review this PR on my side & pass this to the rest of my team to test.

File: validator_client/src/duties_service.rs [hunk 2]
```diff
@@ -42,6 +42,9 @@ const HISTORICAL_DUTIES_EPOCHS: u64 = 2;
 /// At start-up selection proofs will be computed with less lookahead out of necessity.
 const SELECTION_PROOF_SLOT_LOOKAHEAD: u64 = 8;
 
+/// The attestation selection proof lookahead for those running with the --dvt flag.
```
Comment (xenowits) [inline validator_client/src/duties_service.rs]:
```suggestion
/// The attestation selection proof lookahead for those running with the --distributed flag.
```

File: validator_client/src/duties_service.rs [hunk 3]
```diff
@@ -997,7 +1005,13 @@ async fn fill_in_selection_proofs<T: SlotClock + 'static, E: EthSpec>(
                 continue;
             };
 
-            let lookahead_slot = current_slot + SELECTION_PROOF_SLOT_LOOKAHEAD;
+            let selection_look_ahead = if duties_service.distributed {
```
Comment (xenowits) [inline validator_client/src/duties_service.rs]: nit: 
```suggestion
            let selection_lookahead = if duties_service.distributed {
```

**Comment by xenowits** [comment]
Hey @michaelsproul , I have reviewed it & the changes look solid. We'll be running this branch internally on one of our clusters & test it.

**Comment by michaelsproul** [comment]
@xenowits Great to hear! Let us know how it goes, and whether you need any other tweaks

**Comment by dB2510** [comment]
Hey @michaelsproul @AgeManning 👋 

We tested these changes with our clusters and aggregations don't work as it seems like lighthouse VC is not calling selection endpoints to get aggregated selection proof as described in #4248. Will the changes to call those endpoints be included in a separate pull request, since this PR does not query those selection endpoints?

I am referring to these endpoints:
- https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Validator/submitBeaconCommitteeSelections
- https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Validator/submitSyncCommitteeSelections

Thanks

**Comment by AgeManning** [comment]
Hey @dB2510 

Yep, these need to still be added. I recently saw a talk about running obol which indicated the lodestar vc was required to run an obol node. Do you know which other clients support these endpoints, is it just lodestar?

They are in the specification, I'll have a chat with the others about the best path forward from here. Supporting these endpoints will be a future PR.

**Comment by michaelsproul** [comment]
What about the builder registration overrides (https://github.com/sigp/lighthouse/pull/4306), do you still need that too?

It would be good to have a complete list of features required by Obol so we can know the fill scope of changes.

**Comment by michaelsproul** [comment]
@xenowits Confirmed via DM that this PR is a good step forward for Obol support, and that we should merge it to allow further testing.

The other features desired for Obol DVT are:

- https://github.com/sigp/lighthouse/issues/4248
- https://github.com/sigp/lighthouse/pull/5010 (nice to have, prevents error logs)

I'm still double-checking whether we need (update: we do):

- https://github.com/sigp/lighthouse/pull/4306

**Review by dapplion — APPROVED** [review]
Adding the `--distributed` flag is good step, this PR can be merged as is. All logic is correct.

Just merged unstable to fix conflicts

**Comment by michaelsproul** [comment]
I've merged `unstable` again and updated the CLI tests. Let's merge once CI passes

**Comment by michaelsproul** [comment]
@mergifyio queue

**Comment by michaelsproul** [comment]
@mergifyio requeue

**Comment by michaelsproul** [comment]
@mergifyio requeue

**Comment by OisinKyne** [comment]
Hey @AgeManning, appreciate the chat last week. Here’s a synopsis of where I believe we are that (though you wrote the PR so likely have the specifics). 

>   “it seems like lighthouse VC is not calling selection endpoints to get aggregated selection proof as described in #4248. Will the changes to call those endpoints be included in a separate pull request, since this PR does not query those selection endpoints?”  

[This comment above](https://github.com/sigp/lighthouse/pull/4867#issuecomment-1856029402) is the last key update for us on your PR. [This](https://github.com/sigp/lighthouse/pull/4306) was the other thing that’s not that important and is now closed. (So Michaels comment further down of: “I'm still double-checking whether we need (update: we do):” is no longer the case). 

Calling those endpoints (attest aggregate, sync aggregate) during the flow and using the returned selection proof rather than the one generated locally to hash and determine if you’re an aggregator should be what’s left IIUC. [Here is](https://docs.google.com/document/d/1q9jOTPcYQa-3L8luRvQJ-M0eegtba4Nmon3dpO79TMk/edit?usp=sharing) our original implementation note, and [here](https://github.com/ObolNetwork/kurtosis-charon) is a kurtosis repo that might help with testing, though we’ll also happily test once you have a PR or an `unstable` build :)

Let me know if I can answer any other follow ups. Cheers.

**Pull request merged**